guides/howtos/Many to many and upserts.md

```diff
@@ -0,0 +1,207 @@
+    |> Ecto.Changeset.cast(params, [:title, :body])
+    |> Ecto.Changeset.put_assoc(:tags, parse_tags(params))
+  end
+
+  defp parse_tags(params)  do
+    (params["tags"] || "")
+    |> String.split(",")
+    |> Enum.map(&String.trim/1)
+    |> Enum.reject(& &1 == "")
+    |> Enum.map(&get_or_insert_tag/1)
+  end
+
+  defp get_or_insert_tag(name) do
+    Repo.get_by(MyApp.Tag, name: name) ||
+      Repo.insert!(MyApp.Tag, %Tag{name: name})
+  end
+end
+```
+
+In the changeset function above, we moved all the handling of tags to a separate function, called `parse_tags/1`, which checks for the parameter, breaks each tag apart via `String.split/2`, then removes any left over whitespace with `String.trim/1`, rejects any empty string and finally checks if the tag exists in the database or not, creating one in case none exists.
+
+The `parse_tags/1` function is going to return a list of `MyApp.Tag` structs which are then passed to `put_assoc/4`. By calling `put_assoc/4`, we are telling Ecto those should be the tags associated to the post from now on. In case a previous tag was associated to the post and not given in `put_assoc/4`, Ecto will invoke the behaviour defined in the `:on_replace` option, which we have set to `:delete`. The `:delete` behaviour will remove the association between the post and the removed tag from the database.
+
+And that's all we need to use `many_to_many` associations with `put_assoc/4`. `put_assoc/4` is very useful when we want to have more explicit control over our associations and it also works with `has_many`, `belongs_to` and all others association types.
+
+However, our code is not yet ready for production. Let's see why.
+
+## Constraints and race conditions
+
+Remember we added a unique index to the tag `:name` column when creating the tags table. We did so to protect us from having duplicate tags in the database.
+
+By adding the unique index and then using `get_by`  with a `insert!` to get or insert a tag, we introduced a potential error in our application. If two posts are submitted at the same time with a similar tag, there is a chance we will check if the tag exists at the same time, leading both submissions to believe there is no such tag in the database. When that happens, only one of the submissions will succeed while the other one will fail. That's a race condition: your code will error from time to time, only when certain conditions are met. And those conditions are time sensitive.
+
+Many developers have a tendency to think such errors won't happen in practice or, if they happened, they would be irrelevant. But in practice they often lead to very frustrating user experiences. I have heard a first-hand example coming from a mobile game company. In the game, a player is able to play quests and on every quest you have to choose a guest character from another player out of a short list to go on the quest with you. At the end of the quest, you have the option to add the guest character as a friend.
```

Comment:
I would remove this "story". Let's say "Luckily Ecto gives us a mechanism to handle constraint errors from the database." here and skip to the next section.